pkg/kv/kvserver/batcheval/split_trigger_datadriven.go

```diff
@@ -0,0 +1,461 @@
+				desc := rs.desc
+				db := storage.NewDefaultInMemForTesting()
+				defer db.Close()
+				batch := db.NewBatch()
+				defer batch.Close()
+
+				rec := (&MockEvalCtx{
+					ClusterSettings:        tc.st,
+					Desc:                   &desc,
+					Clock:                  tc.clock,
+					AbortSpan:              rs.abortspan,
+					LastReplicaGCTimestamp: rs.lastGCTimestamp,
+					RangeLeaseDuration:     tc.rangeLeaseDuration,
+				}).EvalContext()
+
+				// Write the range state that will be consulted and copied
+				// during the split.
+				sl := stateloader.Make(desc.RangeID)
+				err := sl.SetLease(ctx, batch, nil, rs.lease)
+				require.NoError(t, err)
+				err = sl.SetGCThreshold(ctx, batch, nil, &rs.gcThreshold)
+				require.NoError(t, err)
+				err = sl.SetGCHint(ctx, batch, nil, &rs.gcHint)
+				require.NoError(t, err)
+				err = sl.SetVersion(ctx, batch, nil, &rs.version)
+				require.NoError(t, err)
+
+				// Run the split trigger.
+				_, _, err = splitTrigger(ctx, rec, batch /* bothDeltaMS */, enginepb.MVCCStats{}, split, hlc.Timestamp{})
+				if err != nil {
+					return fmt.Sprintf("splitTrigger error: %v", err)
+				}
+
+				// Update the test context's notion of range state after split.
+				tc.updatePostSplitRangeState(t, ctx, batch, rangeID, split)
+
+				// Print the state of the batch (all keys/values written).
+				var sb strings.Builder
+				iter, err := batch.NewEngineIterator(ctx, storage.IterOptions{
+					UpperBound: roachpb.KeyMax,
+					KeyTypes:   storage.IterKeyTypePointsAndRanges,
+				})
+				require.NoError(t, err)
+
+				defer iter.Close()
+				var lines []string
+				valid, err := iter.SeekEngineKeyGE(storage.EngineKey{Key: roachpb.KeyMin})
+				require.NoError(t, err)
+				for valid {
+					hasPoint, hasRange := iter.HasPointAndRange()
+					if hasPoint {
+						k, err := iter.EngineKey()
+						require.NoError(t, err)
+						v, err := iter.UnsafeValue()
+						if err != nil {
+							return fmt.Sprintf("error getting value: %v", err)
+						}
+						lines = append(lines, fmt.Sprintf("point %s = %x", k, v))
+					}
+					if hasRange {
+						bounds, err := iter.EngineRangeBounds()
+						require.NoError(t, err)
+
+						v := iter.EngineRangeKeys()
+						for i := range v {
+							lines = append(lines, fmt.Sprintf(
+								"range [%s,%s) = %x/%x",
+								bounds.Key, bounds.EndKey, v[i].Value, v[i].Version))
+						}
+					}
+					valid, err = iter.NextEngineKey()
+					require.NoError(t, err)
+				}
+
+				for _, line := range lines {
+					sb.WriteString(line)
+					sb.WriteString("\n")
+				}
+				return sb.String()
+
+			default:
+				return fmt.Sprintf("unknown command: %s", d.Cmd)
+			}
+		})
+	})
+}
+
+type rangeState struct {
+	desc            roachpb.RangeDescriptor
+	lease           roachpb.Lease
+	gcThreshold     hlc.Timestamp
+	gcHint          roachpb.GCHint
+	version         roachpb.Version
+	abortspan       *abortspan.AbortSpan
+	lastGCTimestamp hlc.Timestamp
+}
```

Comment:
Do we need to mirror this in memory and the back-and-forth with storage? Could we only have stuff in storage?